### PR TITLE
WIP: Improve Proofs specs

### PR DESCRIPTION
**Pull request by nicola**

### CODE REVIEW

File: client-data.md [hunk 1]
```diff
@@ -1,21 +1,17 @@
 # Filecoin Client Data Processing
 
-Data that clients wish to store must be pre-processed before it can be sent to miners and have a storage deal made for it.
+Client must pre-process their data into a valid formats before sending them to miners, creating storage deals and generating mining proofs.
```
Comment (nikkolasg) [inline client-data.md]: This way this sentence is written sounds like a client is generating a mining proof, which I don't believe is true, is it ?

File: client-data.md [hunk 2]
```diff
@@ -1,21 +1,17 @@
 # Filecoin Client Data Processing
```
Comment (nikkolasg) [inline client-data.md]: Do we have to repeat filecoin in a subsection of the... filecoin specs ? My philosophy is to get to the point: the less words the better (less cognitive load)

File: client-data.md [hunk 3]
```diff
@@ -1,21 +1,17 @@
...
-1. **Transfer Encoding**
-2. **Storage Encoding**
+- **Transfer Format**
+- **Storage Format**
 
-Transfer Encoding is a pre-processing for the file to allow it to be transferred over the network more easily. Both for the initial transfer to the storage miner, and for later retrievals. This mechanism can be done in any way that is supported by the client and the miner (alternative strategies may be implemented).
+## Transfer Format
 
-Storage Encoding is a pre-processing for the file that allows the Filecoin storage market and proofs to reference the data. Filecoin operates on a special type of merkle tree generated with a function we call `RepHash`. The root hash of this merkle tree is called the `Piece Commitment` or `CommP` and is included in the deal when it is proposed to a miner.
+The Transfer Format is the format to transfer a file over the network. This format is used for the initial transfer (from clients to storage miners) and for later retrievals (from storage miners to the clients).
```
Comment (nikkolasg) [inline client-data.md]: (1) I quite preferred encoding since that is exactly what you describe but I imagine you changed to format because of collision with namings in the proofs or stg ?
(2) Is "Transfer Format" really a "proper name" ? simply have it like "transfer format" seems simpler (i.e. having it a proper name doesn't bring any benefit imo)

File: client-data.md [hunk 4]
```diff
@@ -1,21 +1,17 @@
...
 
-The next sections will go into some additional detail on how each of these encodings works.
-
-## Transfer Encoding
-
-Transfer Encoding may be computed in a variety of ways, as long as it is supported by both the client and miner involved. Here, we document the `unixfsv1` transfer encoding.
+The default Transfer Format is `unixfsv1`, however clients and miner might agree to use other formats of their preference.
```
Comment (nikkolasg) [inline client-data.md]: > however clients and miner might agree to use other formats of their preference.

That is a troubling sentence to be found in a spec. 
(1) I'd rather remove it and leave the implementer without any choices to make (which IS the goal of a spec imo)
(2) IF we want to keep it, then we should use common terminology such as "clients and miner MAY agree to use..." so the understand is clear (it's the same words used in officials RFC and there's even an RFC for that  ;) )

File: client-data.md [hunk 5]
```diff
@@ -27,17 +23,42 @@ The default transfer encoding is Unixfsv1 with the following parameters:
 
 For details on how UnixfsV1 works, see its spec [here](https://github.com/ipfs/specs/tree/master/unixfs).
 
-## Storage Encoding
+## Storage Format
+
+The Storage Format it's the required format for generating Filecoin proofs and hashing sectors data.
```
Comment (nikkolasg) [inline client-data.md]: This is not a sentence. Maybe rephrase it to 
`The Storage Format is the required format for generating Filecoin proofs and hashing sectors data. `

And again I would not put it as a proper name, not sure it adds real value to the table (it's not a product nor a feature etc)

File: proofs.md [hunk 6]
```diff
@@ -1,44 +1,55 @@
...
+The commitments are used to verify that the correct __*original data*__ was sealed, and that the correct __*sealed data*__ is the subject of later __*Proofs of Spacetime*__, proving that this data is being stored continuously.
 
-`Seal` operates by performing a slow encoding of the __*unsealed sector*__ — such that it is infeasible for a dishonest prover to computationally regenerate the __*sealed sector*__ quickly enough to satisfy subsequent required __*Proofs of Space-Time*__ — thus ensuring that the __*sealed sector*__ remains manifest as a unique, concrete representation of the __*original data*__.
+`Seal` operates by performing a slow encoding of the __*unsealed sector*__ — such that it is infeasible for a dishonest prover to computationally regenerate the __*sealed sector*__ quickly enough to satisfy subsequent required __*Proofs of Spacetime*__ — thus ensuring that the __*sealed sector*__ remains manifest as a unique, concrete representation of the __*original data*__.
```
Comment (nikkolasg) [inline proofs.md]: `quickly enough to satisfy subsequent required ` => `quickly enough to produce subsequent required` 
I don't believe we can say "satisfy a proof", do we ? (but we can say the proof satisfy the verification checks or stg)

File: proofs.md [hunk 7]
```diff
@@ -61,12 +72,11 @@ Seal
 
 ```
 
-### VerifySeal
+### PoRep.VerifySeal
```
Comment (nikkolasg) [inline proofs.md]: Same comment on PoRep than before

File: proofs.md [hunk 8]
```diff
@@ -1,44 +1,55 @@
...
+- __*prover:*__ the party who generates a proof, in Filecoin it's always the Storage Miner.
+- __*verifier:*__ the party who verifies a proof generated by a __*prover*__, in Filecoin it's a full node.
+- __*sectors count:*__ the number of sectors over which a proof-of-spacetime is performed (`POST_SECTORS_COUNT`).
+
+## Proof-of-Replication Algorithms
 
 __*Filecoin Proof of Replication*__ generates a unique copy (__*sealed sector*__) of a __*sector's*__ __*original data*__, a __*Multi-SNARK proof*__, and a set of __*commitments*__ identifying the __*sealed sector*__ and linking it to the corresponding __*unsealed sector*__.
 
-## Seal
+### PoRep.Seal
```
Comment (nikkolasg) [inline proofs.md]: I'd remove "PoRep" in the title since (1) we are already in the proof of replication section and (2) PoRep is not defined before.

It'd probably be easier to read `Seal` as a title and just before having a sentence that says that PoRep means Proof of Replication.

Comment (nikkolasg) [inline proofs.md]: I'd remove "PoRep" in the title since (1) we are already in the proof of replication section and (2) PoRep is not defined before.

It'd probably be easier to read `Seal` as a title and just before having a sentence that says that PoRep means Proof of Replication.

File: proofs.md (hunk 7 above)
Comment (nikkolasg) [reply proofs.md]: As well, below is just "Seal" and this is good !!

File: proofs.md [hunk 9]
```diff
@@ -141,15 +155,17 @@ VerifyPoSt
...
+ ) err Error | 
+  isValid bool                     // true iff the provided Proof of Spacetime is valid
 ```
 
----
+------
+
 ## Piece Inclusion Proof
 
 ### PieceInclusionProof
```
Comment (nikkolasg) [inline proofs.md]: Feels like we could just inline that text since this subsection has the same name as its parent section "Piece Inclusion Proof"

File: client-data.md (hunk 2 above)
Comment (nicola) [reply client-data.md]: ack

File: client-data.md (hunk 1 above)
Comment (nicola) [reply client-data.md]: yup, it's also redundant comment, removing it

File: client-data.md (hunk 4 above)
Comment (nicola) [reply client-data.md]: option 2 is best

File: client-data.md [hunk 10]
```diff
@@ -27,17 +23,42 @@ The default transfer encoding is Unixfsv1 with the following parameters:
...
+- **Fr32 Padding**: Every 32 bytes blocks must contain two zeroes in the most significant bits (every 254 bits must be followed by 2 zeroes if interpreted as little-endian number). That is, for each block, `0x11000000 & block[31] == 0`.
+- **Piece Padding**: blocks of 32 zero bytes are added so that the total number of blocks (including *piece padding*) is a power of two.
+
+**Why do we need a special Storage Encoding Format?** In the Filecoin proofs we do operations in an arithmetic field of size `p`, where `p` is a prime of size `2^255`, hence the size of the data blocks must be smaller than `p`. We cautiously decide to have data blocks of size 254 to avoid possible overflows (data blocks numerical representation is bigger than `p`). 
+
 
-Storage Encoding is computed by building a merkle tree out of the data using `RepHash`.
 
-Inputs to `RepHash` must first be preprocessed then padded.
+## Piece Commitment
```
[the remark this answers is not in the document]
Comment (nicola) [reply client-data.md]: because it's not necessarely a file, it could be part of a file as you are saying. In addition, there are some proofs called Piece Inclusion Proofs, which would be File Inclusion Proof and we would like to keep FIP for Filecoin Improvement Process

File: proofs.md [hunk 11]
```diff
@@ -1,44 +1,55 @@
...
+## Overview
 
-- __*Proof of Replication*__ proves that a unique copy of a given __*sector*__ has been created. The `Seal` operation creates this unique copy and generates a corresponding __*Proof of Replication*__.
-- __*Proof of Space-Time*__ proves that an arbitrary number of __*sealed sectors*__ existed over a specified period of time in their own dedicated storage — as opposed to being generated on-the-fly at proof time.
-- __*Piece Inclusion Proof*__ proves that a given __*piece*__ is contained within a specified __*sealed sector*__.
-- __*Proof of Retrievability*__ is a merkle proof that a given challenged leaf is present in an extant __*sealed
-sector*__.
+The Filecoin protocol uses cryptographic proofs to ensure the two following features:
+
+- **Storage Based Consensus**: Miners' power in the consensus is proportional to their amount of storage. Miner increase their power by collecting users' data in sectors
```
[the remark this answers is not in the document]
Comment (nicola) [reply proofs.md]: ack

File: proofs.md [hunk 12]
```diff
@@ -1,44 +1,55 @@
...
 
-The Filecoin protocol requires a means of generating and verifying the following cryptographic proofs:
+## Overview
 
-- __*Proof of Replication*__ proves that a unique copy of a given __*sector*__ has been created. The `Seal` operation creates this unique copy and generates a corresponding __*Proof of Replication*__.
-- __*Proof of Space-Time*__ proves that an arbitrary number of __*sealed sectors*__ existed over a specified period of time in their own dedicated storage — as opposed to being generated on-the-fly at proof time.
-- __*Piece Inclusion Proof*__ proves that a given __*piece*__ is contained within a specified __*sealed sector*__.
-- __*Proof of Retrievability*__ is a merkle proof that a given challenged leaf is present in an extant __*sealed
-sector*__.
+The Filecoin protocol uses cryptographic proofs to ensure the two following features:
```
[the remark this answers is not in the document]
Comment (nicola) [reply proofs.md]: ack

File: proofs.md [hunk 13]
```diff
@@ -1,44 +1,55 @@
...
+
+- **Storage Based Consensus**: Miners' power in the consensus is proportional to their amount of storage. Miner increase their power by collecting users' data in sectors
+- **Verifiable Storage Market**: Miners must be proving that they are dedicating unique physical space for each copy of the clients data through a period of time.
+
+We briefly describe the three main Filecoin proofs:
+
+- __*Proof-of-Replication (PoRep)*__ proves that a Storage Miner is dedicating unique dedicated storage for each ***sector***. Filecoin Storage Miners collect new clients' data in a sector, run a slow encoding process (called `Seal`) and generate a proof (`SealProof`) that the encoding was generated correctly.
+
+  In Filecoin, PoRep provides two guarantees: (1) *space-hardness*: Storage Miners cannot lie about the amount of space they are dedicating to Filecoin in order to gain more power in the consensus; (2) *replication*: Storage Miners are dedicating unique storage for each copy of their clients data.
```
[the remark this answers is not in the document]
Comment (nicola) [reply proofs.md]: what do you mean?

File: proofs.md [hunk 14]
```diff
@@ -1,44 +1,55 @@
...
+
+We briefly describe the three main Filecoin proofs:
+
+- __*Proof-of-Replication (PoRep)*__ proves that a Storage Miner is dedicating unique dedicated storage for each ***sector***. Filecoin Storage Miners collect new clients' data in a sector, run a slow encoding process (called `Seal`) and generate a proof (`SealProof`) that the encoding was generated correctly.
+
+  In Filecoin, PoRep provides two guarantees: (1) *space-hardness*: Storage Miners cannot lie about the amount of space they are dedicating to Filecoin in order to gain more power in the consensus; (2) *replication*: Storage Miners are dedicating unique storage for each copy of their clients data. 
+
+- __*Proof of Spacetime*__ proves that an arbitrary number of __*sealed sectors*__ existed over a specified period of time in their own dedicated storage — as opposed to being generated on-the-fly at proof generation time.
+
+- __*Piece-Inclusion-Proof*__ proves that a given __*piece*__ is contained within a specified __*sealed sector*__.
```
[the remark this answers is not in the document]
Comment (nicola) [reply proofs.md]: it's the same

File: proofs.md (hunk 8 above)
Comment (nicola) [reply proofs.md]: ack

File: zigzag-porep.md [hunk 15]
```diff
@@ -1,85 +1,131 @@
...
 
-- Calculate `ReplicaID` using `Hash` (Blake2s):
-```
-ReplicaID := Hash(ProverID || SectorID || ticket)
-```
-- Perform `RepHash` on `Data` to yield `CommD` and `TreeD`:
-```
-CommD, TreeD = RepHash(data)
-```
+The SNARK circuit proves that given a Merkle root `CommD`, `CommRLast`, and `commRStar`, that the prover knew the correct replicated data at each layer.
```
[the remark this answers is not in the document]
Comment (nicola) [reply zigzag-porep.md]: ack

File: zigzag-porep.md [hunk 16]
```diff
@@ -1,85 +1,131 @@
...
+
+In Filecoin, PoRep provides two guarantees: (1) *space-hardness*: Storage Miners cannot lie about the amount of space they are dedicating to Filecoin in order to gain more power in the consensus; (2) *replication*: Storage Miners are dedicating unique storage for each copy of their clients data. 
+
+Glossary:
+
+- __*sector:*__ a fixed-size block of data of `SECTOR_SIZE` bytes which generally contains clients' data.
+- __*unsealed sector:*__ a concrete representation (on disk or in memory) of a sector's that follows the "Storage Format" described in [Filecoin Client Data Processing](client-data.md#storage-format) (currently `paddedfr32v1` is the required default).
+- __*sealed sector:*__  a concrete representation (on disk or in memory) of the unique replica generated by `Seal` from an __*unsealed sector*__. 
+- __*piece:*__ a block of data of at most `SECTOR_SIZE` bytes which is generally is a client's file or part of.
```
[the remark this answers is not in the document]
Comment (nicola) [reply zigzag-porep.md]: resolving

File: zigzag-porep.md [hunk 17]
```diff
@@ -1,85 +1,131 @@
...
+### Public Parameters
+
+The following public parameters are used in the ZigZag Replication and Proof Generation algorithms:
+
+TODO: the Appendix should explain why we picked those values
+
+
+| name | type | description | value |
+| --- | --- | --- | ---: |
+| `SECTOR_SIZE` | `uint` | Number of nodes in the DRG. | `64GiB` |
```
[the remark this answers is not in the document]
Comment (nicola) [reply zigzag-porep.md]: ack

File: zigzag-porep.md [hunk 18]
```diff
@@ -1,85 +1,131 @@
...
+| --- | --- | --- | ---: | ---- |
+| `PARENTS_COUNT` | `uint` | Total number of parent nodes |`EXPANSION_DEGREE + BASE_DEGREE` | `13` |
+| `GRAPH_SIZE` | `uint` | Number of nodes in the graph  | `SECTOR_SIZE / NODE_SIZE` | `2,147,483,648` |
+| `TREE_DEPTH` | `uint` | Height of the Merkle Tree of a sector | `LOG_2(GRAPH_SIZE)` | `31` |
+
+
+The following additional public parameters are required:
+
+- `TAPER` : Float: Fraction of each layer's challenges by which to reduce next-lowest layer's challenge count.
+- `TAPER_LAYERS`: uint: Number of layers
```
[the remark this answers is not in the document]
Comment (nicola) [reply zigzag-porep.md]: this should be put in a table anyway. will do this for now

File: zigzag-porep.md [hunk 19]
```diff
@@ -1,85 +1,131 @@
...
+*ZigZag* is a specific Proof-of-Replication construction that we use in Filecoin. ZigZag has been designed by [Ben Fisch at EUROCRYPT19](https://eprint.iacr.org/2018/702.pdf).  In high level, ZigZag ensures that the *Replicate* step is a slow non-parallelizable sequential process by using some special type of graphs called Depth Robust Graphs (we refer to as "DRG").
 
-## Replication
+**Encoding using DRGs**. By positioning data blocks into nodes in the DRG, we sequentially encode each node in the graph using its encoded parents. DRG graphs ensure that this process is not likely to be parallelizable.
 
-#### Public Parameters
+**Layering DRGs**. ZigZag repeates this encoding by layering DRG graphs `LAYERS` times. The data represented in each DRG layer is the data encoded in the previous layer. We connect different layers using Bipartite Expander Graphs and at each layer, we reverse the graph edges with a technique which we call zigzag. The combination of DRGs, expander graphs and zigzag guarantee the security property of PoRep. The final encoded layer is the final replica
```
[the remark this answers is not in the document]
Comment (nicola) [reply zigzag-porep.md]: ack

File: zigzag-porep.md [hunk 20]
```diff
@@ -299,77 +286,55 @@ func baseParents(node uint) (parents [BASE_DEGREE]uint) {
...
-        if graphIsReversed() {
-            assert(parent >= rawNode)
-        } else {
-            assert(parent <= rawNode)
-        }
+	feistelKeys := []uint{1, 2, 3, 4} // TODO
+  
+	for i := 0, p := 0; i < EXPANSION_DEGREE; i++ {
+		a := node * EXPANSION_DEGREE + i
+    transformed := feistelPermute(GRAPH_SIZE * EXPANSION_DEGREE, a, feistelKeys)
```
Comment (schomatis) [inline zigzag-porep.md]: nit: tabulation

File: zigzag-porep.md [hunk 21]
```diff
@@ -528,5 +650,41 @@ return layer_proofs, CommRstar, CommRLast
 
 TODO: reconcile outputs of non-circuit proof with inputs to circuit proof.
 
-## Circuit Proof Generation
+### SNARK Proof Generation
+
 See [# ZigZag: Offline PoRep Circuit Spec](zigzag-circuit.md) for details of Circuit Proof Generation.
+
+## Appendex
```
Comment (schomatis) [inline zigzag-porep.md]: nit: appendix

**Comment by schomatis** [comment]
Also, it's getting harder and harder to review the spec with so many changes on top of changes (GitHub code review flow doesn't help either), so I'd suggest merging what we have now (with the corresponding fixes) and pushing additional content to a new PR.

**Comment by porcuquine** [comment]
I tend to agree with @schomatis. I propose @nicola addresses whichever outstanding comments he thinks are worth incorporating. Once that is done, I would like to make another pass over it. I can clean up a lot of prose while doing that.

**Comment by nicola** [comment]
merging now, but more work needed.
